### PR TITLE
Symbols procs

### PR DESCRIPTION
**Pull request by 0sc**

### CODE REVIEW

File: symbol.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
+# SYMBOLS#PROCS
+
+Thank you for accepting this mission, and the mission is nothing else but to add Symbols#procs to our Ruby toolchain . However, to achieve our mission, we need to have expertise of  **symbols** & **Procs** and then with their combined power we will take on the mission.
+
+If you have some difficulties with any of these concepts, do not worry, I made arrangements for you. [The Ruby_Newbie Guide to Symbols](http://www.troubleshooters.com/codecorn/ruby/symbols.htm), has exactly what you will need for symbols and [Ruby Explained: Blocks, Procs, and Lambdas, aka "Closures"](http://www.eriktrautman.com/posts/ruby-explained-blocks-procs-and-lambdas-aka-closures). Once you are well equipped, onward to the mission.
```
Comment (ghost) [inline symbol.md]: I am not sure this statement is complete.
`I made arrangements for you. The Ruby_Newbie Guide to Symbols, has exactly what you will need for symbols and Ruby Explained: Blocks, Procs, and Lambdas, aka "Closures".`
Can you rephrase it?

File: symbol.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+Let's construct a class for our soldier object
```
Comment (ghost) [inline symbol.md]: Why do we need a soldier object? Can you clarify here?

File: symbol.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+    def grade_a_weapon?
+      weapon.length > 5
+    end
+  end
+
+  soldier = Soldier.new("John Doe", "AK-47")
+
+  # From our symbol armory
+  soldier.send(:name) #=> John Doe
```
Comment (ghost) [inline symbol.md]: Is there any point using `send` here? If there is no specific reason, you should use the dot notation instead.

File: symbol.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+Next up, we'll create a collection of all soldiers that signed up for the mission
+
+```ruby
+  soldiers = []
+  soldiers << Soldier.new('Big Mike', "Bazooka")
+  soldiers << Soldier.new('Small John', "Pistol")
+  soldiers << Soldier.new('Lightening Larry', "Uzi")
+
+  # From our proc armory
+  proc = Proc.new { |soldier| puts "Hail #{soldier.name}"}
```
Comment (ghost) [inline symbol.md]: Maybe, you can start with a normal block:

``` ruby
soldiers.each do |soldier|
  puts "Hail #{soldier.name}"
end
```

Before using a proc which is passed as a block.

File: symbol.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+  #=> Hail Big Mike
+  #=> Hail Small John
+  #=> Hail Lightening Larry
+
+```
+
+What if we need a collection of just the names of the soldiers, we could using the map function do
+
+```ruby
+  soldiers.map{ |soldier|  soldier.send(:name) } #=> ["Big Mike", "Small John", "Lightening Larry"]
```
Comment (ghost) [inline symbol.md]: Again, why use the `send` method. When writing ruby code, we should avoid using needless metaprogramming.

File: symbol.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+  soldiers.map(&proc) #=> ["Big Mike", "Small John", "Lightening Larry"]
+```
+
+With symbol#proc, we can simplify this further
+
+```ruby
+  soldiers.map(&:name) #=> ["Big Mike", "Small John", "Lightening Larry"]
+```
+
+Now you know why we needed to make those stops we made at symbols and procs as such the question now won't be what each components of the code mean but rather, **How they work together?**
```
Comment (ghost) [inline symbol.md]: You missed `and` after `procs`. Maybe you can just end the sentence there before continuing with `As such`.

File: symbol.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+_Here's one for you to try out: write code to retrieve the last_name of soldiers with grade_a_weapon. Share your implementation in the comment section._
+
+Now let's match towards a more ambitious adventure.
+
+We can override the `to_proc` method in the Symbol class to achieve custom result. For instance say we want to always append the team captain's name to all `to_proc` requests;
+
+```ruby
+  class Symbol
+    def to_proc
+      Proc.new { |x| "#{x.send(self)} (signed John Doe Commander)" }
```
Comment (ghost) [inline symbol.md]: Inform the readers of the dark side of doing this. They should never do it. This is for the purpose of learning.

File: symbol.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+Symbol to proc are really useful and often come in handy in the drive for cleaner concise codes. One rule of thumb for when to use symbol to proc is whenever you find yourself having blocks whose sole function is call a single method on the passed arg i.e
+```ruby
+  something {|i| i.foo }
+
+  # should be replaced with
+  something(&:foo)
+```
+examples
+```ruby
+  (1..100).reject{|num| num.odd? }
```
Comment (ghost) [inline symbol.md:181]: Can you show your readers the alternative implementation for each of this example using `Symbols#to_proc`?

File: symbol.md [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+  (1..3).map {|num| num.to_s}
+  (1..100).inject { |a, b| a + b}
+  words.map { |s| s.length }
+  %w(john paul ringo george).map { |p| p.capitalize }
+  collection.select { |x| x.valid }
+  soldiers.map{|x| x.last_name }
+  soldiers.map{|x| x.first_name }
+```
+Congrats you've successfully completed the mission. Show your mastery of `symbol#proc` by rewriting the examples above using what you just learned; paste your implementation in the comment section below.
```
Comment (ghost) [inline symbol.md]: :+1: Very good read. It was `mission impossible` but I made it back. All hail `Symbols#to_proc`.

File: symbol.md [hunk 10]
```diff
@@ -0,0 +1,181 @@
+# SYMBOLS#PROCS
```
Comment (ghost) [inline symbol.md]: Generally, can you have defined subsections?

File: symbol.md (hunk 1 above)
Comment (0sc) [inline symbol.md]: Very right, you are. Thanks

File: symbol.md (hunk 8 above)
Comment (0sc) [inline symbol.md:181]: My thought was to leave it to them to practice with and then engage in the comment section.

File: symbol.md (hunk 3 above)
Comment (0sc) [inline symbol.md]: The reason for it's use here (and hereafter) is to make users comfortable with using the `send` method. I was trying to avoid introducing them _sharply_ to it's use in the implementation of `:to_proc`, which might have me explaining with examples what the `:send` method does.

File: symbol.md (hunk 10 above)
Comment (0sc) [inline symbol.md]: How do you mean?

Comment (ydaniju) [inline symbol.md]: Jaw breaking!

Comment (ghost) [inline symbol.md]: @andela-ooranagwa Let me explain in detail. Let's assume the title of the blog post is `The Mystery behind &:symbol(Symbol#to_proc)`. This post can have several subsections eg
- Mission Impossible => Introduction
- The Soldier Class
- The Army => soldier collection
- Shorthand Notation: Symbol#to_proc
- Overriding Symbol#to_proc
- It's All about to_proc method: Custom class
- Conclusion.

This is just an example, the subsections you may come up with can be different.

File: symbol.md (hunk 4 above)
Comment (ghost) [inline symbol.md]: You can point the reader to a different blog post eg [This blog post](http://goodheads.io/2016/03/12/ruby-on-steroids-the-magic-of-metaprogramming-method-spells/) on goodheads for an explanation on send.
